### ISSUE TITLE
CV 的主要问题

### ISSUE THREAD

**Issue by JiaxiangBU** [JiaxiangBU/tutoring#77]
### 简述问题

采用了K-fold 过多，最后在验证集上面的产生了过拟合？

### 过程中产生的报错 (可选)

...

遇到报错的时候，推荐采用 debugging 思路，定位具体报错信息和报错的函数或者对象，具体可以参考例子
https://jiaxiangbu.github.io/learn_rstudioapi/analysis/rstudio-and-api-learning-notes.html#section-6.2

在书写 Rmd 文档时，

1. `library` 放在前面，先让第三方包 import，这样可以让后面的函数顺利执行。
1. 导入数据用相对路径，避免其他使用者需要重新修改路径。如果使用`here`函数，`?here::here` 查看使用方法
1. 不要重复定义函数，已经定义，使用 `source` 调用。
1. 文档编码是编码 `GBK`的，修改为 `UTF-8`，方法是在 RStudio 点击左上角的 <kbd>File</kbd>，点击 <kbd>Save with Encoding</kbd>，选择 `UTF-8`
1. 代码注意缩进，或者 <kbd>ctrl</kbd> + <kbd>shift</kbd> + <kbd>A</kbd>  reformat 代码
1. 很多 bug 都是书写不规范导致的，先 debug 一下看看。

### 有关联的问题

直接把链接复制在此。

### 复现代码和数据

首次提问，可以建立自己全拼文件夹，见 https://github.com/JiaxiangBU/tutoring
在当前文件夹下，

把反馈的**截图**代码和输入数据上传，我在你的代码上进行修改。

1. 数据传  `output` 文件夹
1. 代码 (`.Rmd`, `.md`) 文档传 `analysis` 文件夹
    1. `.md` 可以展示目前代码状况，不需要我执行代码就可以看到结果，方便快速查看问题。
1. 命名的规范如，`190101-who_am_i-name.postfix`

我们接下来反馈的时候给上有复现的代码和数据，这样我可以很快给你 debug。

1. 相关的 commit 回复到 对应的 issue，这样方便之后跟踪每个问题
1. 一个模块一个模块的解决，基本上一个问题，建立一个独立的issue，不同文档归属到对应的issue
1. 使用 [GitHub Markdown](https://guides.github.com/features/mastering-markdown/) 代码方式，保证 issue 的整洁

代码尽量避免截图，因为无法复制截图复现，因此更好的方式复制代码。

如果是 R 代码，按照以下输入

```r
...
````

如果是 SQL 代码，按照以下输入

```sql
...
```

### Todos

1. [ ] assign 给对应 owner
1. [ ] 打上对应标签
1. [ ] 问题解决后记得 close

**Comment by JiaxiangBU** [comment]
```BibTex
@article{doi:10.1002/cem.1310,
author = {Esbensen, Kim H. and Geladi, Paul},
title = {Principles of Proper Validation: use and abuse of re-sampling for validation},
journal = {Journal of Chemometrics},
volume = {24},
number = {3‐4},
pages = {168-187},
keywords = {Principles of Proper Validation (PPV), future performance assessment, test set validation, cross-validation, re-sampling, predictive regression, Theory of Sampling (TOS)},
doi = {10.1002/cem.1310},
url = {https://onlinelibrary.wiley.com/doi/abs/10.1002/cem.1310},
eprint = {https://onlinelibrary.wiley.com/doi/pdf/10.1002/cem.1310},
abstract = {Abstract Validation in chemometrics is presented using the exemplar context of multivariate calibration/prediction. A phenomenological analysis of common validation practices in data analysis and chemometrics leads to formulation of a set of generic Principles of Proper Validation (PPV), which is based on a set of characterizing distinctions: (i) Validation cannot be understood by focusing on the methods of validation only; validation must be based on full knowledge of the underlying definitions, objectives, methods, effects and consequences—which are all outlined and discussed here. (ii) Analysis of proper validation objectives implies that there is one valid paradigm only: test set validation. (iii) Contrary to much contemporary chemometric practices (and validation myths), cross-validation is shown to be unjustified in the form of monolithic application of a one-for-all procedure (segmented cross-validation) on all data sets. Within its own design and scope, cross-validation is in reality a sub-optimal simulation of test set validation, crippled by a critical sampling variance omission, as it manifestly is based on one data set only (training data set). Other re-sampling validation methods are shown to suffer from the same deficiencies. The PPV are universal and can be applied to all situations in which the assessment of performance is desired: prediction-, classification-, time series forecasting-, modeling validation. The key element of PPV is the Theory of Sampling (TOS), which allow insight into all variance generating factors, especially the so-called incorrect sampling errors, which, if not properly eliminated, are responsible for a fatal inconstant sampling bias, for which no statistical correction is possible. In the light of TOS it is shown how a second data set (test set, validation set) is critically necessary for the inclusion of the sampling errors incurred in all ‘future’ situations in which the validated model must perform. Logically, therefore, all one data set re-sampling approaches for validation, especially cross-validation and leverage-corrected validation, should be terminated, or at the very least used only with full scientific understanding and disclosure of their detrimental variance omissions and consequences. Regarding PLS-regression, an emphatic call is made for stringent commitment to test set validation based on graphical inspection of pertinent t–u plots for optimal understanding of the X–Y interrelationships and for validation guidance. QSAR/QSAP forms a partial exemption from the present test set imperative with no generalization potential. Copyright © 2010 John Wiley \& Sons, Ltd.},
year = {2010}
}
```

BibTex 如上，可以参考这篇论文。

**Comment by JiaxiangBU** [comment]
Esbensen and Geladi (2010) 主要讨论的问题是样本迁移 (data drift)。举一个例子。

1. 假设有 1000 个样本，2:8 分为 train 和 test 组。
1. train 使用 k-fold 进行 cv，hand-out 每次用 800*0.8 训练 (training)，800*0.2 验证 (validation)，因此一共可以进行 5-fold。
1. 如果在第一个 fold 中，validation 已经出现了样本迁移，那么这个5-fold cv 的模型即使误差很小，在 test 组表现不好的。
1. train上表现好 + test 组上表现不好 = 过拟合。

**Comment by JiaxiangBU** [comment]
> Within its own design and scope, cross‐validation is in reality a
> sub‐optimal simulation of test set validation, **crippled** by a
> critical sampling variance omission, as it **manifestly** is based on
> one data set only (training data set). Other re‐sampling validation
> methods are shown to suffer from the same deficiencies. (Esbensen and
> Geladi 2010)

和我的想法一样，就是说 k-fold 始终是局部最优，是 train 组的 bias 和 variance 最小化，而非总体 (train 和
test) 一起的最小化。 Esbensen and Geladi (2010) 说的 one data set only (training
data set) 和 deficiencies 就是这个意思。

<div id="refs" class="references">

<div id="ref-Esbensen2010">

Esbensen, Kim H., and Paul Geladi. 2010. “Principles of Proper
Validation: Use and Abuse of Re-Sampling for Validation.” *Journal of
Chemometrics* 24 (3‐4): 168–87. <https://doi.org/10.1002/cem.1310>.

</div>

</div>

**Comment by JiaxiangBU** [comment]
> Drift means that e.g. an instrument's response/true calibration changes slowly over time. So the generalization error for unknown cases may not be the same as for unknown future cases. You arrive at instructions like "redo calibration daily/weekly/..." if you find drift during validation, but this needs test sets systematically acquired later than the training data.

样本迁移的意思就是目标变量 y 的的变化随着时间变化**比较快**.

**Comment by JiaxiangBU** [comment]
另外最好的方式，是做一个测试，你可以直接用 `ggplot2::diamonds`这个数据集进行测试，你会发现他们效果不是严格满足下面关系的。

perf(k=10000) > perf(k=1000) > perf(k=100) > perf(k=10)

数据集大小如下

``` r
nrow(ggplot2::diamonds)
#> [1] 53940
```

<sup>Created on 2019-08-09 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by JiaxiangBU** [comment]
CV  不是单调的，我之前上课的时候，记得是这么说的。

只能说，

1. 不做 CV < 做 CV
1. k = 2 < k = 10
1. k = 10 ??? k = 1000 这个不一定。
